File: crates/prolly/bindings/wasm/examples/diff-merge.ts
import assert from "node:assert/strict";
import { existsSync, readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

import { loadProllyWasm } from "../src/index.ts";

const here = dirname(fileURLToPath(import.meta.url));
const pkgWasm = join(here, "../pkg/prolly_wasm_bg.wasm");
if (!existsSync(pkgWasm)) throw new Error("Run `npm --prefix crates/prolly/bindings/wasm run build:wasm` first.");

const wasm = await loadProllyWasm("../pkg/prolly_wasm.js", readFileSync(pkgWasm));
const bytes = (value: string): Uint8Array => new TextEncoder().encode(value);
const text = (value: Uint8Array | null | undefined): string | null =>
  value == null ? null : new TextDecoder().decode(value);

const engine = wasm.WasmProllyEngine.memory();
let base = engine.create();
base = engine.put(base, bytes("doc:title"), bytes("Draft"));

const left = engine.put(base, bytes("doc:body"), bytes("Hello"));
const right = engine.put(base, bytes("doc:tags"), bytes("example"));

const leftChanges = engine.diff(base, left);
assert.equal(leftChanges.length, 1);

const merged = engine.merge(base, left, right, "prefer_right");
assert.equal(text(engine.get(merged, bytes("doc:body"))), "Hello");
assert.equal(text(engine.get(merged, bytes("doc:tags"))), "example");

console.log(`diff_merge: merged ${leftChanges.length} left-side change`);

File: crates/prolly/examples/diff_merge.rs
use prolly::{Config, Diff, MemStore, Prolly};

fn main() -> Result<(), prolly::Error> {
    let prolly = Prolly::new(MemStore::new(), Config::default());

    let base = prolly.create();
    let base = prolly.put(&base, b"doc:title".to_vec(), b"Draft".to_vec())?;

    let left = prolly.put(&base, b"doc:body".to_vec(), b"Hello".to_vec())?;
    let right = prolly.put(&base, b"doc:tags".to_vec(), b"example".to_vec())?;

    let left_changes = prolly.diff(&base, &left)?;
    assert!(matches!(
        left_changes.as_slice(),
        [Diff::Added { key, val }] if key == b"doc:body" && val == b"Hello"
    ));

    let merged = prolly.merge(&base, &left, &right, None)?;
    assert_eq!(prolly.get(&merged, b"doc:body")?, Some(b"Hello".to_vec()));
    assert_eq!(prolly.get(&merged, b"doc:tags")?, Some(b"example".to_vec()));

    println!("merged {} left-side change", left_changes.len());
    Ok(())
}
